client/asset/dcr/dcr.go

```diff
@@ -5425,6 +5540,261 @@ func (dcr *ExchangeWallet) createSig(tx *wire.MsgTx, idx int, pkScript []byte, a
+// checks all the pending transactions to see if they have been mined into
+// a block, and if so, updates the transaction history to reflect the
+// block height.
+func (dcr *ExchangeWallet) checkPendingTxs(ctx context.Context, tip uint64) {
+	txHistoryDB := dcr.txDB()
+	if txHistoryDB == nil {
+		return
+	}
+
+	synced, _, err := dcr.SyncStatus()
+	if err != nil {
+		dcr.log.Errorf("Error getting sync status: %v", err)
+		return
+	}
+	if !synced {
+		dcr.log.Infof("not synced")
+		return
+	}
+
+	// First, check to see if there are any recieving transactions we have not
+	// yet seen.
+	{
+		const blockQueryBuffer = 3
+		var blockToQuery uint64
+		lastQuery := dcr.receiveTxLastQuery.Load()
+
+		if lastQuery != 0 && lastQuery < tip-blockQueryBuffer {
+			blockToQuery = lastQuery - blockQueryBuffer
+		} else if lastQuery >= tip-blockQueryBuffer {
+			blockToQuery = tip - blockQueryBuffer
+		}
+
+		recentTxs, err := dcr.wallet.ListSinceBlock(ctx, int32(blockToQuery), int32(tip), int32(tip))
+		if err != nil {
+			dcr.log.Errorf("Error listing transactions since block %d: %v", blockToQuery, err)
+			recentTxs = nil
+		} else {
+			dcr.receiveTxLastQuery.Store(tip)
+			err = txHistoryDB.SetLastReceiveTxQuery(tip)
+			if err != nil {
+				dcr.log.Errorf("Error setting last query to %d: %v", tip, err)
+			}
+		}
+
+		addToHistoryIfNew := func(tx *walletjson.ListTransactionsResult, txType asset.TransactionType) {
+			fmt.Printf("~~~~~~~ tx cat %v type %+v\n", tx.Category, *tx.TxType)
+			txHash, err := chainhash.NewHashFromStr(tx.TxID)
+			if err != nil {
+				dcr.log.Errorf("Error decoding txid %s: %v", tx.TxID, err)
+				return
+			}
+			_, err = txHistoryDB.GetTx(txHash.String())
+			if err == nil {
+				return
+			}
+			if !errors.Is(err, asset.CoinNotFoundError) {
+				dcr.log.Errorf("Error getting tx %s: %v", tx.TxID, err)
+				return
+			}
+
+			var fee uint64
+			if tx.Fee != nil {
+				// Fee always seems to be negative in btcwallet, but just
+				// in case.
+				if *tx.Fee < 0 {
+					fee = toAtoms(-*tx.Fee)
+				} else {
+					fee = toAtoms(*tx.Fee)
+				}
+			}
+
+			dcr.addTxToHistory(txType, txHash, toAtoms(tx.Amount), fee, nil, nil, true)
+		}
+
+		for _, tx := range recentTxs {
+			if *tx.TxType == walletjson.LTTTRegular && tx.Category == "receive" {
+				addToHistoryIfNew(&tx, asset.Receive)
+			} else if *tx.TxType == walletjson.LTTTVote {
+				addToHistoryIfNew(&tx, asset.TicketVote)
+			} else if *tx.TxType == walletjson.LTTTRevocation {
+				addToHistoryIfNew(&tx, asset.TicketRevokation)
+			}
+		}
+	}
+
+	// Not just the map must be copied here, but the ExtendedWalletTx
+	// as well.
+	pendingTxsCopy := make(map[chainhash.Hash]btc.ExtendedWalletTx, len(dcr.pendingTxs))
+	dcr.pendingTxsMtx.RLock()
+	for hash, tx := range dcr.pendingTxs {
+		pendingTxsCopy[hash] = *tx
+	}
+	dcr.pendingTxsMtx.RUnlock()
+
+	handlePendingTx := func(txHash chainhash.Hash, tx *btc.ExtendedWalletTx) {
```

Comment:
Why not just do everything from here down under lock?

Reply:
I didn't want it to hold the `pendingTxsMtx` for too long. It would hold up every other function that submits a transaction.